doc: clarify stream errors while reading and writing (#29653)
nodejs/node: doc/api/stream.md

#### Errors While Writing

Errors occurring during the processing of the [`writable._write()`][],

> [`writable._writev()`][] and [`writable._final()`] methods should be propagated

Reviewer: I would make this stronger with a `must` rather than `should`
```suggestion
[`writable._writev()`][] and [`writable._final()`] methods must be propagated
```

Reviewer: and likewise below...

Author: I like that. However, the reason I used `should` is so that stream users don't 100% assume this is the case... since the ecosystem might not conform... a typical example of such cases https://github.com/nodejs/node/issues/29699.

Given that you still think we should change I'm all for it.

Author: No you are right. No question about it.

#### Errors While Reading

> Errors occuring during processing of the [`readable._read()`][] should be

Reviewer: ```suggestion
Errors occurring during processing of the [`readable._read()`][] must be
```

propagated through the [`readable.destroy(err)`][readable-_destroy] method.
Throwing an `Error` from within [`readable._read()`][] or manually emitting an
`'error'` event results in undefined behavior.

```js
const { Readable } = require('stream');

const myReadable = new Readable({
  read(size) {
    const err = checkSomeErrorCondition();
    if (err) {

>       this.destory(err);

Reviewer 2: Destroy

>       this.destory(err);

Author: ```suggestion
      this.destroy(err);
```

[`process.stdout`]: process.html#process_process_stdout
[`readable.push('')`]: #stream_readable_push
[`readable.setEncoding()`]: #stream_readable_setencoding_encoding

> [`readable._read()`]: stream.html#stream_readable_read_size_1

Reviewer 3: ```suggestion
[`readable._read()`]: #stream_readable_read_size_1
```

Reviewer 3: Also: as bottom references are sorted in ASCII order, the undercore needs to be placed before the lower case letters.

[`writable.cork()`]: #stream_writable_cork
[`writable.end()`]: #stream_writable_end_chunk_encoding_callback
[`writable.uncork()`]: #stream_writable_uncork

> [`writable._final()`]: #stream.html#stream_writable_final_callback

Reviewer 3: ```suggestion
[`writable._final()`]: #stream.html#stream_writable_final_callback
```

> [`writable._write()`]: #stream.html#stream_writable_write_chunk_encoding_callback_1

Reviewer 3: ```suggestion
[`writable._write()`]: #stream_writable_write_chunk_encoding_callback_1
```

> [`writable._writev()`]: #stream.html#stream_writable_writev_chunks_callback

Reviewer 3: ```suggestion
[`writable._writev()`]: #stream_writable_writev_chunks_callback
```
